tools/bot/matrix.go

```diff
@@ -0,0 +1,147 @@
+package bot
+
+import (
+	"TUM-Live/tools"
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	"strconv"
+)
+
+// Matrix strategy
+type Matrix struct {
+}
+
+// Message represents a Matrix message event that includes html formatting as specified
+// in https://spec.matrix.org/v1.2/client-server-api/#mroommessage-msgtypes
+type Message struct {
+	MsgType       string `json:"msgtype"`
+	Body          string `json:"body"`
+	Format        string `json:"format"`
+	FormattedBody string `json:"formatted_body"`
+}
+
+// Login represents a login request that is used for authentication as specified
+// in https://spec.matrix.org/v1.2/client-server-api/#login
+type Login struct {
+	Type     string `json:"type"`
+	User     string `json:"user"`
+	Password string `json:"password"`
+}
+
+// MessageResponse is the response for a message in our case
+type MessageResponse struct {
+	EventID string `json:"event_id"`
+}
+
+// LoginResponse is the response for a login request
+type LoginResponse struct {
+	UserId      string `json:"user_id"`
+	AccessToken string `json:"access_token"`
+	HomeServer  string `json:"home_server"`
+	DeviceId    string `json:"device_id"`
+	WellKnown   struct {
+		MHomeserver struct {
+			BaseUrl string `json:"base_url"`
+		} `json:"m.homeserver"`
+	} `json:"well_known"`
+}
+
+var (
+	// Base Links
+	clientUrl = "https://" + tools.Cfg.Alerts.Matrix.Homeserver + "/_matrix/client/r0/"
+	loginUrl  = clientUrl + "login"
+)
+
+const (
+	// Maximum transaction ID
+	maxID = 10000
+
+	// Suffixes and Prefixes for sending a message to a room
+	accessTokenSuffix = "?access_token="
+	roomSuffix        = "rooms/"
+	roomMsgPrefix     = "/send/m.room.message/"
+
+	// Constants describing the messages itself
+	msgType     = "m.text"
+	msgFormat   = "org.matrix.custom.html"
+	loginMethod = "m.login.password"
+
+	contentType = "application/json"
+)
+
+// SendBotMessage sends a formatted message to a matrix room
+func (ma *Matrix) SendBotMessage(info InfoMessage) error {
+	id := strconv.Itoa(rand.Intn(maxID)) // transaction id
+	authToken, err := getAuthToken()
+	if err != nil {
+		return err
+	}
+	if authToken == "" {
+		return errors.New("authentication failed, could not get token")
+	}
+
+	var roomMessageSuffix = roomMsgPrefix + id + accessTokenSuffix
+	url := clientUrl + roomSuffix + tools.Cfg.Alerts.Matrix.RoomID + roomMessageSuffix + authToken
+	matrixMessage := Message{
+		MsgType:       msgType,
+		Body:          generateInfoText(info),
+		Format:        msgFormat,
+		FormattedBody: getFormattedMessageText(info),
+	}
+	matrixMessageJSON, err := json.Marshal(matrixMessage)
+	if err != nil {
+		return err
+	}
+	err = sendMessageRequest(url, bytes.NewBuffer(matrixMessageJSON))
+	return err
+}
+
+// sendMessageRequest sends a PUT request to url with a given body
+func sendMessageRequest(url string, body io.Reader) error {
```

Comment:
Can this be a method on the matrix type to abstract it out of the package root? 

```suggestion
func (m Matrix)sendMessageRequest(url string, body io.Reader) error {
```